orchagent/bfdorch.cpp

```diff
@@ -551,3 +615,116 @@ uint32_t BfdOrch::bfd_src_port(void)
     return (port++);
 }
 
+void BfdOrch::notify_session_state_down(const string& key)
+{
+    size_t found_vrf = key.find(delimiter);
+    if (found_vrf == string::npos)
+    {
+        SWSS_LOG_ERROR("Failed to parse key %s, no vrf is given", key.c_str());
+        return;
+    }
+
+    size_t found_ifname = key.find(delimiter, found_vrf + 1);
+    if (found_ifname == string::npos)
+    {
+        SWSS_LOG_ERROR("Failed to parse key %s, no ifname is given", key.c_str());
+        return;
+    }
+    string vrf_name = key.substr(0, found_vrf);
+    string alias = key.substr(found_vrf + 1, found_ifname - found_vrf - 1);
+    IpAddress peer_address(key.substr(found_ifname + 1));
+    BfdUpdate update;
+    update.peer = get_state_db_key(vrf_name, alias, peer_address);
+    update.state = SAI_BFD_SESSION_STATE_DOWN;
+    notify(SUBJECT_TYPE_BFD_SESSION_STATE_CHANGE, static_cast<void *>(&update));
+    return;
+}
+
+void BfdOrch::handleTsaStateChange(bool tsaState)
+{
+    SWSS_LOG_INFO("BfdOrch TSA state Changed to %d.\n", int(tsaState));
+    tsa_enabled = tsaState;
+    for (auto it : bfd_session_cache)
+    {
+        if (tsaState == true)
+        {
+            notify_session_state_down(it.first);
+            if (!remove_bfd_session(it.first))
+            {
+                SWSS_LOG_ERROR("Failed to remove BFD session %s\n", it.first.c_str());
+            }
+        } else
+        {
+            if (!create_bfd_session(it.first, it.second))
+            {
+                SWSS_LOG_ERROR("Failed to create BFD session %s\n", it.first.c_str());
+            }
+        }
+    }
+}
+
+BgpGlobalStateOrch::BgpGlobalStateOrch(DBConnector *db, string tableName):
+    Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+    tsa_enabled = false;
+    SWSS_LOG_ERROR("BgpGlobalStateOrch init complete\n");
+}
+
+BgpGlobalStateOrch::~BgpGlobalStateOrch(void)
+{
+    SWSS_LOG_ENTER();
+}
+
+bool BgpGlobalStateOrch::getTsaState()
+{
+    SWSS_LOG_ENTER();
+    return tsa_enabled;
+}
+void BgpGlobalStateOrch::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string key =  kfvKey(t);
+        string op = kfvOp(t);
+        auto data = kfvFieldsValues(t);
+
+        if (op == SET_COMMAND)
+        {
+            for (auto i : data)
+            {
+                auto value = fvValue(i);
+                auto type = fvField(i);
+                SWSS_LOG_INFO("SET on key %s, data T %s, V %s\n", key.c_str(), type.c_str(), value.c_str());
+                if (type == "tsa_enabled")
+                {
+                    bool state = true ? value == "true" : false;
+                    if (tsa_enabled != state)
```

Comment:
This whole section can be handled in `BfdOrch::handleTsaStateChange`

Reply:
done. moved the whole thing to BgpGlobalStateOrch

Reply:
Not sure I understand. Can you please check my original comment?

Reply:
as discussed in the review, the BgpGlobalStateOrch now handles and keeps the state of tsa_enabled.